pkg/registry/core/service/ipallocator/controller/repairip.go

```diff
@@ -0,0 +1,539 @@
+
+	err := r.syncService(eKey.(string))
+	r.handleSvcErr(err, eKey)
+
+	return true
+}
+
+func (r *RepairIPAddress) handleSvcErr(err error, key interface{}) {
+	if err == nil {
+		r.svcQueue.Forget(key)
+		return
+	}
+
+	if r.svcQueue.NumRequeues(key) < maxRetries {
+		klog.V(2).InfoS("Error syncing Service, retrying", "service", key, "err", err)
+		r.svcQueue.AddRateLimited(key)
+		return
+	}
+
+	klog.Warningf("Dropping Service %q out of the queue: %v", key, err)
+	r.svcQueue.Forget(key)
+	runtime.HandleError(err)
+}
+
+// syncServices reconcile the Service ClusterIPs to verify that each one has the corresponding IPAddress object associated
+func (r *RepairIPAddress) syncService(key string) error {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	svc, err := r.serviceLister.Services(namespace).Get(name)
+	if err != nil {
+		// nothing to do
+		return nil
+	}
+	if !helper.IsServiceIPSet(svc) {
+		// didn't need a ClusterIP
+		return nil
+	}
+
+	for _, clusterIP := range svc.Spec.ClusterIPs {
+		ip := netutils.ParseIPSloppy(clusterIP)
+		if ip == nil {
+			// ClusterIP is corrupt, ClusterIPs are already validated, but double checking here
+			// in case there are some inconsistencies with the parsers
+			r.recorder.Eventf(svc, nil, v1.EventTypeWarning, "ClusterIPNotValid", "ClusterIPValidation", "Cluster IP %s is not a valid IP; please recreate Service", ip)
+			runtime.HandleError(fmt.Errorf("the ClusterIP %s for Service %s/%s is not a valid IP; please recreate Service", ip, svc.Name, svc.Namespace))
+			continue
+		}
+
+		family := getFamilyByIP(ip)
+		if network, ok := r.networkByFamily[family]; !ok {
+			// this service is using an IPFamily no longer configured on cluster
+			r.recorder.Eventf(svc, nil, v1.EventTypeWarning, "ClusterIPNotValid", "ClusterIPValidation", "Cluster IP %s(%s) is of ip family that is no longer configured on cluster; please recreate Service", ip, family)
+			runtime.HandleError(fmt.Errorf("the ClusterIP [%v]: %s for Service %s/%s is of ip family that is no longer configured on cluster; please recreate Service", family, ip, svc.Name, svc.Namespace))
+			continue
+		} else if !network.Contains(ip) {
+			// ClusterIP is out of range
+			r.recorder.Eventf(svc, nil, v1.EventTypeWarning, "ClusterIPOutOfRange", "ClusterIPAllocation", "Cluster IP [%v]: %s is not within the configured Service CIDR %s; please recreate service", family, ip, network.String())
+			runtime.HandleError(fmt.Errorf("the ClusterIP [%v]: %s for Service %s/%s is not within the service CIDR %s; please recreate", family, ip, svc.Name, svc.Namespace, network.String()))
+			continue
+		}
+
+		ipaddress, err := r.ipAddressLister.Get(ip.String())
+		if apierrors.IsNotFound(err) {
+			// ClusterIP doesn't seem to be allocated,
+			r.recorder.Eventf(svc, nil, v1.EventTypeWarning, "ClusterIPNotAllocated", "ClusterIPAllocation", "Cluster IP [%v]: %s is not allocated; repairing", family, ip)
```

Comment:
This is a 3 states state-machine
1. Was not allocated: service ip was selected but have not correctly create the ipaddress object.
2. Fixed-Allocated: we managed to get the ip the service said it wants
3. Failed-Allocated: can't allocate the ip the service wants (only if we get duplicate error on create). 
4. Failed-general error: api-server was not cooperating at the time of create.